feat(connlib): create flow on ICMP error "prohibited"

thomaseizinger opened the pull request:
In Firezone, a Client requests an "access authorization" for a Resource on the fly when it sees the first packet for said Resource going through the tunnel. If we don't have a connection to the Gateway yet, this is also where we will establish a connection and create the WireGuard tunnel.

In order for this to work, the access authorization state between the Client and the Gateway MUST NOT get out of sync. If the Client thinks it has access to a Resource, it will just route the traffic to the Gateway. If the access authorization on the Gateway has expired or vanished otherwise, the packets will be black-holed.

Starting with #9816, the Gateway sends ICMP errors back to the application whenever it filters a packet. This can happen either because the access authorization is gone or because the traffic wasn't allowed by the specific filter rules on the Resource.

With this patch, the Client will attempt to create a new flow (i.e. re-authorize) traffic for this resource whenever it sees such an ICMP error, therefore acting as a way of synchronizing the view of the world between Client and Gateway should they ever run out of sync.

Testing turned out to be a bit tricky. If we let the authorization on the Gateway lapse naturally, we portal will also toggle the Resource off and on on the Client, resulting in "flushing" the current authorizations. Additionally, it the Client had only access to one Resource, then the Gateway will gracefully close the connection, also resulting in the Client creating a new flow for the next packet.

To actually trigger this new behaviour we need to:

- Access at least two resources via the same Gateway
- Directly send `reject_access` to the Gateway for this particular resource

To achieve this, we dynamically eval some code on the API node and instruct the Gateway channel to send `reject_access`. The connection stays intact because there is still another active access authorization but packets for the other resource are answered with ICMP errors.

To achieve a safe roll-out, the new behaviour is feature-flagged. In order to still test it, we now also allow feature flags to be set via env variables.

Resolves: #10074

thomaseizinger committed:
Add feature-flag

thomaseizinger committed:
Check for filter prohibited

thomaseizinger committed:
Create better ctor functions

thomaseizinger committed:
Send create flow on filter prohibited icmp error

thomaseizinger committed:
Log trigger for connection intent

thomaseizinger committed:
Add integration test

thomaseizinger committed:
Expose `reject_access` directly

thomaseizinger committed:
Use a 2nd resource

thomaseizinger committed:
Allow setting feature flags from env

thomaseizinger committed:
Ignore error to first resource

thomaseizinger committed:
Set feature flag in docker

thomaseizinger commented on elixir/apps/api/lib/api/gateway/channel.ex:
```diff
@@ -351,6 +351,15 @@ defmodule API.Gateway.Channel do
     end
   end
 
+  # Helper to directly send reject_access in integration tests
+  def handle_info(
+        {{:reject_access, gateway_id}, client_id, resource_id},
+        %{assigns: %{gateway: %{id: gateway_id}}} = socket
+      ) do
+    push(socket, "reject_access", %{client_id: client_id, resource_id: resource_id})
+    {:noreply, socket}
+  end
```
I ended up adding this to reliably send `reject_access` in the integration test.

jamilbk replied:
This will simplify the call site a bit.

```suggestion
  def handle_info({:send_test_reject_access, client_id, resource_id}, socket) do
    push(socket, "reject_access", %{client_id: client_id, resource_id: resource_id})
  end
  
  def send_test_reject_access(resource_name) do
    [client_id | _] =
      Domain.Clients.Client
      |> Repo.all()
      |> Domain.Clients.preload_clients_presence()
      |> Enum.filter(&(&1.online?))
    resource =
      Domain.Resources.Resource
      |> Domain.Repo.all()
      |> Enum.filter(&(&1.name == resource_name))
      
    send(self(), {:send_test_reject_access, client_id, resource.id})
  end
```

thomaseizinger commented on rust/telemetry/src/feature_flags.rs:
```diff
@@ -241,6 +263,44 @@ impl FeatureFlags {
+    }
+}
+
+fn update_from_env(flags: FeatureFlagsResponse) -> FeatureFlagsResponse {
+    FeatureFlagsResponse {
+        icmp_unreachable_instead_of_nat64: env_or(
+            "FZFF_ICMP_UNREACHABLE_INSTEAD_OF_NAT64",
+            flags.icmp_unreachable_instead_of_nat64,
+        ),
+        drop_llmnr_nxdomain_responses: env_or(
+            "FZFF_DROP_LLMNR_NXDOMAIN_RESPONSES",
+            flags.drop_llmnr_nxdomain_responses,
+        ),
+        stream_logs: env_or("FZFF_stream_logs", flags.stream_logs),
+        map_enobufs_to_wouldblock: env_or(
+            "FZFF_MAP_ENOBUFS_TO_WOULDBLOCK",
+            flags.map_enobufs_to_wouldblock,
+        ),
+        gateway_userspace_dns_a_aaaa_records: env_or(
+            "FZFF_GATEWAY_USERSPACE_DNS_A_AAAA_RECORDS",
+            flags.gateway_userspace_dns_a_aaaa_records,
+        ),
+        icmp_error_unreachable_prohibited_create_new_flow: env_or(
+            "FZFF_ICMP_ERROR_UNREACHABLE_PROHIBITED_CREATE_NEW_FLOW",
+            flags.icmp_error_unreachable_prohibited_create_new_flow,
+        ),
+    }
+}
```
We can now set feature flags via env variables.

klochowicz replied:
nice addition, I imagine it will become useful for other things.

thomaseizinger commented on scripts/tests/create-flow-from-icmp-error.sh:
```diff
@@ -0,0 +1,22 @@
+# Try to access resource 1 again
+# First one for each IP will fail because we get an ICMP error.
+client_curl_resource "172.20.0.100/get" || true
+client_curl_resource "[172:20:0::100]/get" || true
```
The nature of ICMP errors is that they will still fail the initial connection but the subsequent ones will go through without issues.

klochowicz replied:
this code doesn't seem to check that accessing the resource failed, it only silences the error

I'll throw in a small commit that deals with this, ideally something like this (will expand our integration library) 

`expect_error client_curl_resource "172.20.0.100/get"`

klochowicz replied:
@thomaseizinger I've pushed a commit that uses the new functionality, along with some unit test coverage of the helper & adding it to CI. The latter is a bit awkward as I didn't want to create a whole new stage of CI, so I piggybacked it into static analysis part of the workflows. Feel free to move it. :)

thomaseizinger replied:
Thanks! I hope you don't mind that I've removed those tests again. I don't think we need maintain an entire bash script to test a 5 line bash function. It seems out of balance in terms of the amount of code we have in here vs the risk that these tests eliminate.

Bash is notorious for how bad it is and how many pitfalls it has. Testing bash with bash doesn't seem like a good idea because I now need to review even more bash code to make sure it is actually correct.

klochowicz replied:
all good;  I'll add a dedicated PR for these, with proper argumentation. I agree that we shouldn't just slip them here. and then we can decide to not merge anyway. 

the main thing is the actual functionality, and I guess it's tested in the integration test.

thomaseizinger commented on rust/connlib/tunnel/src/client.rs:
```diff
@@ -472,6 +474,20 @@ impl ClientState {
             &mut self.udp_dns_sockets_by_upstream_and_query_id,
         );
 
+        if feature_flags::icmp_error_unreachable_prohibited_create_new_flow()
+            && let Ok(Some((failed_packet, error))) = packet.icmp_error()
+            && error.is_unreachable_prohibited()
+            && let Some(resource) = self.get_resource_by_destination(failed_packet.dst())
```
The actual logic here is quite simple :)

thomaseizinger committed:
Use local resource instead

thomaseizinger committed:
Skip integration tests if we don't have minimal version yet

thomaseizinger committed:
Simplify version check

thomaseizinger commented on .github/workflows/_integration_tests.yml:
```diff
@@ -128,6 +130,15 @@ jobs:
       - uses: ./.github/actions/ghcr-docker-login
         with:
           github_token: ${{ secrets.GITHUB_TOKEN }}
+      - name: Check minimum client version
+        id: version_check
+        if: ${{ matrix.test.min_client_version }}
+        continue-on-error: true
+        run: |
+          ACTUAL_VERSION=$(docker run ${{ inputs.client_image }}:${{ inputs.client_tag }} firezone-headless-client --version | awk '{print $2}')
+          MIN_VERSION="${{ matrix.test.min_client_version }}"
+
+          [ "$(printf '%s\n' "$MIN_VERSION" "$ACTUAL_VERSION" | sort --version-sort | head -n1)" == "$MIN_VERSION" ]
```
I added this so we can have this test in here even though it does not pass the compatibility tests (see https://github.com/firezone/firezone/actions/runs/18023609619?pr=10462). With this check in here, we can have the integration test run for the current build of the client and it will also start running as soon as we make a release.

jamilbk commented on scripts/tests/lib.sh:
```diff
@@ -33,6 +33,23 @@ function client_nslookup() {
+    local site_name="$1"
+    local resource_name="$2"
+
+    docker compose exec -T api bin/api rpc "
+Application.ensure_all_started(:domain)
+account_id = \"c89bcc8c-9392-4dae-a40d-888aef6d28e0\"
+
+[gateway_group] = Domain.Gateways.Group.Query.not_deleted() |> Domain.Gateways.Group.Query.by_account_id(account_id) |> Domain.Gateways.Group.Query.by_name(\"$site_name\") |> Domain.Repo.all()
+[gateway_id | _] = Domain.Gateways.Presence.Group.list(gateway_group.id) |> Map.keys()
+[client_id | _] = Domain.Clients.Presence.Account.list(account_id) |> Map.keys()
+[resource] = Domain.Resources.Resource.Query.not_deleted() |> Domain.Resources.Resource.Query.by_account_id(account_id) |> Domain.Repo.all() |> Enum.filter(&(&1.name == \"$resource_name\"))
+
+Domain.PubSub.Account.broadcast(account_id, {{:reject_access, gateway_id}, client_id, resource.id})
+"
```
This is quite a lot of inline Elixir that's not statically checked. How about defining a simple helper function in the Gateway module and just calling that helper function from here?

E.g.



```suggestion
    docker compose exec -T api bin/api rpc "API.Gateway.Channel.send_test_reject_access(\"$resource_name\")"
```

thomaseizinger replied:
> This is quite a lot of inline Elixir that's not statically checked. How about defining a simple helper function in the Gateway module and just calling that helper function from here?

I have no opinion in regards to that. I originally built it like this because it was quicker to iterate on inline Elixir code vs having to recompile constantly.

thomaseizinger replied:
Feel free to push a commit!

thomaseizinger replied:
I do think we want to pass resource and site name unless you know a way how to target the correct gateway.

jamilbk reviewed (commented):
Ok I gave this some thought. I think there are two cases we send this "Prohibited by filter" response right?

1. Whenever the resource has been removed for some reason in the Gateway's peer state but has _not_ been reset on the Client, or some other kind of state drift / missed update. In this case we do want to trigger a new intent/flow I think because it will effectively heal the situation.
2. Whenever traffic filters don't allow talking to a resource the client is trying to talk to. Question: does this happen for IP stack mismatches? For example, if an HTTP application on the client uses happy eyeballs to try connections to certain DNS resources over and over again that have `ipv4_only` specified, would that create a bunch of useless "flows" (authorizations)?

For (1) above, I would think we should retry because we can reasonably expect the situation to heal.

For (2), this will never heal no matter how many times a new flow is authorized.

Perhaps we need two different ICMP responses to represent these two cases? Or am I thinking about this wrong?



Also left a suggestion on how to simplify the RPC call site.

jamilbk commented on rust/connlib/tunnel/src/client.rs:
```diff
@@ -472,6 +474,20 @@ impl ClientState {
             &mut self.udp_dns_sockets_by_upstream_and_query_id,
         );
 
+        if feature_flags::icmp_error_unreachable_prohibited_create_new_flow()
```
`new_flow_on_icmp_prohibited`?

thomaseizinger replied:
The name is already used in a bunch of places, including PostHog. I think I'll just keep it as is.

klochowicz committed:
tests: Add ability to expect command erroring

Add necessary infra with unit tests, and use it in
`create-flow-from-icmp-error` test where we can now positively assert
that the first request should fail.

Include new unit tests in CI workflows.

thomaseizinger commented on scripts/tests/unit-test-lib.sh:
I think this is a bit overkill.

thomaseizinger commented on scripts/tests/lib.sh:
```diff
@@ -86,3 +86,33 @@ function create_token_file {
+        echo "ERROR: Expected code $expected_code but got $actual_code: $*"
+        return 1
+    fi
+}
```
This appears to be unused?

klochowicz replied:
yeah, unused in the library, but tested; useful generic component.

I'd imagine you could even use it here, I just didn't know what error code you'd get, and I didn't want to assume!

thomaseizinger committed:
Remove shell unit tests

klochowicz reviewed (approved):
LGTM (I'm happy with addressing Elixir suggestion in a follow-up if it's easier)
